File: 2021/flareon2021/02_known/a.py
import struct

def __ROL__(n, d):
    return ((n << d)|(n >> (8 - d))) & 0xFF

def __ROR__(n, d):
    return ((n >> d)|(n << (8 - d))) & 0xFF

def decrypt_once(data, key):
    result = b""
    for i in range(8):
        result += struct.pack("B", (__ROL__(key[i] ^ data[i], i) - i) & 0xff)
    return result

enc = open("./Files/critical_data.txt.encrypted", "rb").read()

png_magic = b"\x89\x50\x4E\x47\x0D\x0A\x1A\x0A"
png_magic_enc = b"\xC7\xC7\x25\x1D\x63\x0D\xF3\x56"

key = b""
for i in range(8):
    key += struct.pack("B", __ROR__((png_magic[i] + i) & 0xFF, i) ^ png_magic_enc[i])

print("key =", key)

dec = b""
for i in range(0, len(enc), 8):
    dec += decrypt_once(enc[i:i+8], key)

print(dec)

File: 2021/flareon2021/02_known/a.go
package main

import (
	"fmt"
	"math/bits"
	"os"
	"encoding/hex"
)

func decryptOnce(data []uint8, key []uint8) []uint8 {
	var result []uint8
	for i := 0; i < 8; i++ {
		decByte := bits.RotateLeft8(key[i] ^ data[i], i) - uint8(i)
		result = append(result, decByte)
	}
	return result
} 

func main() {
	data, _ := os.ReadFile("./Files/critical_data.txt.encrypted")
	data = []uint8(data)
	pngMagic, _ := hex.DecodeString("89504E470D0A1A0A")
	pngMagic = []uint8(pngMagic)
	pngMagicEnc, _ := hex.DecodeString("C7C7251D630DF356")
	pngMagicEnc = []uint8(pngMagicEnc)

	var key []uint8
	for i := 0; i < 8; i++ {
		keyByte := bits.RotateLeft8(pngMagic[i] + uint8(i), -i) ^ pngMagicEnc[i]
		key = append(key, keyByte)
	}
	fmt.Println(key)

	var flag []uint8
	for i := 0; i < len(data); i += 8 {
		flag = append(flag, decryptOnce(data[i:i+8], key)...)
	}
	fmt.Println(string(flag))
}
